packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -1063,38 +1035,44 @@ function renderRoot(
         // Don't bother with a very short suspense time.
         if (msUntilTimeout > 10) {
           if (workInProgressRootHasPendingPing) {
-            // This render was pinged but we didn't get to restart earlier so try
-            // restarting now instead.
-            prepareFreshStack(root, expirationTime);
-            return renderRoot.bind(null, root, expirationTime);
+            const lastPingedTime = root.lastPingedTime;
+            if (lastPingedTime === NoWork || lastPingedTime >= expirationTime) {
+              // This render was pinged but we didn't get to restart earlier so
+              // try restarting now instead.
+              root.lastPingedTime = expirationTime;
+              prepareFreshStack(root, expirationTime);
+              return;
+            }
           }
 
-          const nextKnownPendingLevel = root.nextKnownPendingLevel;
-          if (nextKnownPendingLevel !== NoWork) {
-            // There's lower priority work. It might be unsuspended. Try rendering
-            // at that level.
-            return renderRoot.bind(null, root, nextKnownPendingLevel);
+          const nextTime = getNextRootExpirationTimeToWorkOn(root);
+          if (nextTime !== NoWork && nextTime !== expirationTime) {
+            // There's additional work on this root.
+            return;
           }
           if (
             lastSuspendedTime !== NoWork &&
             lastSuspendedTime !== expirationTime
           ) {
-            // We should prefer to render the fallback of at the last
-            // suspended level.
-            return renderRoot.bind(null, root, lastSuspendedTime);
+            // We should prefer to render the fallback of at the last suspended
+            // level. Ping the last suspended level to try rendering it again.
+            root.lastPingedTime = lastSuspendedTime;
+            return;
```

Comment:
I'm paranoid about these not having a callback scheduled and us stalling.

Reply:
I put `ensureRootIsScheduled` in the finally block of `performWorkOnRoot` to try to account for this.

I think it could theoretically happen for some of the early flush paths, which bypass `performWorkOnRoot`. But since they bypass `performWorkOnRoot`, they also won’t cause an existing callback to be canceled.